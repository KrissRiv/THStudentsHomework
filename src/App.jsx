import { React, useState } from "react";
import "./App.css";
import { client } from "./services/courses-service";


function App() {
 const [isLoading, setLoading] = useState(false);
 const [searchTerm, setSearchTerm] = useState("");
 const [listItems, setListItems] = useState([]);
 const [listParentItems, setListParentItems] = useState([]);


 const getCourses = () => {
   client
     .get(`?query=${searchTerm}`)
     .then((response) => {
       if (response.status !== 200) {
         throw new Error("Network response was not ok");
       }
       const data = response.data;
       const parents = data.filter((item) => item.parent_id === 0);
       const list = [];
       const restList = [];
       parents.forEach((itemList) => {
         list.push({
           ...itemList,
           childs: data.filter((item) => item.parent_id === itemList.id),
         });
       });
       list.forEach((item) =>
         item.childs.map((child) => {
           restList.push({
             ...child,
             childs: data.filter((item) => item.parent_id === child.id),
           });
         })
       );
       setListItems(restList);
       setListParentItems(list);
     })
     .catch((error) => {
       throw new Error("Network response was not ok");
     })
     .finally(() => setLoading(false));
 };


 const handlerInput = (e) => {
   setSearchTerm(e.target.value);
 };


 const getLastChilds = (id) => {
   return listItems.map(
     (item) =>
       item.childs.length > 0 && (
         <ul>
           {item.childs.map(
             (child) =>
               child.parent_id === id && <li key={child.id}>--{child.name}</li>
           )}
         </ul>
       )
   );
 };


 const getChilds = (childs) => {
   return childs.map((child) => (
     <>
       <li key={child.id}>-{child.name}</li>
       {getLastChilds(child.id)}
     </>
   ));
 };


 const getParentList = () => {
   return listParentItems.map((parent, i) => (
     <>
       <li key={i}>{parent.name}</li>
       {parent.childs.length > 0 && <ul>{getChilds(parent.childs)}</ul>}
     </>
   ));
 };


 const handlerSubmit = (e) => {
   setLoading(true);
   e.preventDefault();
   getCourses();
 };


 return (
   <>
     TH Students Homework
     <hr />
     <form onSubmit={handlerSubmit}>
       <input name="search" onChange={handlerInput} autoFocus data-testid="search" />
       <button type="submit" data-testid="go">
         Go
       </button>
     </form>
     {isLoading && <p>Loading...</p>}
     {listParentItems?.length > 0 && (
       <ul data-testid="courses-collection">{getParentList()}</ul>
     )}
   </>
 );
}


export default App;