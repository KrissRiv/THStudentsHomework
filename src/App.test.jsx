import { render, screen } from '@testing-library/react'

import App from './App'
import { describe } from 'vitest'

describe("App", () => {
  it('Contains heading', () => {
    render(<App />)
    const heading = screen.getByText("TH Students Homework")
    expect(heading).toBeInTheDocument()
  })
  it("Contain input search", () => {
    render(<App />)
    const input = screen.getByTestId("search")
    expect(input).toBeInTheDocument()
  })
  it("Contain button", () => {
    render(<App />)
    const button = screen.getByTestId("go")
    expect(button).toBeInTheDocument()
  })
});
