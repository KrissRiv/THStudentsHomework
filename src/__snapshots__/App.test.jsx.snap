// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > Render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1
        data-testid="title"
      >
        TH Students Homework
      </h1>
      <hr />
      <form>
        <div
          class="input-group mb-3"
        >
          <input
            aria-describedby="button-addon2"
            aria-label="Recipient's username"
            class="form-control"
            data-testid="search"
            name="search"
            placeholder="What do you learn today?"
            type="text"
          />
          <button
            class="btn btn-primary btn-outline-secondary text-white"
            data-testid="go"
            id="button-addon2"
            type="submit"
          >
            Go
          </button>
        </div>
      </form>
    </div>
    <div>
      <h1
        data-testid="title"
      >
        TH Students Homework
      </h1>
      <hr />
      <form>
        <div
          class="input-group mb-3"
        >
          <input
            aria-describedby="button-addon2"
            aria-label="Recipient's username"
            class="form-control"
            data-testid="search"
            name="search"
            placeholder="What do you learn today?"
            type="text"
          />
          <button
            class="btn btn-primary btn-outline-secondary text-white"
            data-testid="go"
            id="button-addon2"
            type="submit"
          >
            Go
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <h1
      data-testid="title"
    >
      TH Students Homework
    </h1>
    <hr />
    <form>
      <div
        class="input-group mb-3"
      >
        <input
          aria-describedby="button-addon2"
          aria-label="Recipient's username"
          class="form-control"
          data-testid="search"
          name="search"
          placeholder="What do you learn today?"
          type="text"
        />
        <button
          class="btn btn-primary btn-outline-secondary text-white"
          data-testid="go"
          id="button-addon2"
          type="submit"
        >
          Go
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
